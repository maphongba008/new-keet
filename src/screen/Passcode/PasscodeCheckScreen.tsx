import React, { memo, useCallback, useEffect } from 'react'
import { Text, TextStyle, View } from 'react-native'

import { IconButton } from 'component/Button'
import { NavBar } from 'component/NavBar'
import SvgIcon from 'component/SvgIcon'
import { colors, createThemedStylesheet } from 'component/theme'
import { UI_SIZE_24 } from 'lib/commonStyles'
import { useAppStore, useTimeout } from 'lib/hooks'
import { getStoragePasscode } from 'lib/localStorage'
import {
  back,
  getCurrentRoute,
  reset,
  SCREEN_HOME,
  SCREEN_PASSCODE_CHECK,
} from 'lib/navigation'

// Adjust the import path as necessary

import { useStrings } from 'i18n/strings'

import { BIOMETRICS_SUCCESS, PasscodeInput } from './PasscodeInput'
import PasscodeMessage from './PasscodeMessage'
import { MessageType, usePasscodeStore } from './usePasscodeStore'

interface PasscodeCheckProps {
  title?: string
  description?: string
  titleStyleProp?: TextStyle
  onSuccess?: () => void
  hasBackButton?: boolean
  hasCloseButton?: boolean
}

const PasscodeCheckScreen: React.FC<PasscodeCheckProps> = ({
  title,
  description,
  titleStyleProp,
  onSuccess,
  hasBackButton = false,
  hasCloseButton = false,
}) => {
  const strings = useStrings()
  const styles = getStyles()
  const { message, messageType, setMessage, setMessageType, resetMessage } =
    usePasscodeStore()

  const handleCorrectPasscodeWithDelay = useTimeout(() => {
    useAppStore.setState({
      isAuthenticated: true,
    })
    if (onSuccess) {
      onSuccess()
    } else {
      const currentRoute = getCurrentRoute()
      if (currentRoute === SCREEN_PASSCODE_CHECK) {
        back()
      } else {
        reset(SCREEN_HOME)
      }
    }
  }, 500)

  const handlePasscodeComplete = useCallback(
    (passcode: string) => {
      const isCorrectPasscode =
        passcode === BIOMETRICS_SUCCESS || passcode === getStoragePasscode()

      if (isCorrectPasscode) {
        setMessageType(MessageType.Success)
        setMessage(strings.passcode.passcodeCorrect)
        handleCorrectPasscodeWithDelay()
      } else {
        setMessageType(MessageType.Error)
        setMessage(strings.passcode.passcodeDoesNotMatch)
      }
    },
    [
      setMessage,
      setMessageType,
      strings.passcode.passcodeCorrect,
      strings.passcode.passcodeDoesNotMatch,
      handleCorrectPasscodeWithDelay,
    ],
  )
  useEffect(() => {
    resetMessage()
  }, [resetMessage])

  return (
    <>
      <NavBar
        title={null}
        left={
          hasBackButton ? (
            <IconButton onPress={back}>
              <SvgIcon
                name="arrowLeft"
                width={UI_SIZE_24}
                height={UI_SIZE_24}
              />
            </IconButton>
          ) : null
        }
        middle={
          !title ? (
            <Text style={styles.title}>{strings.passcode.inputPasscode}</Text>
          ) : null
        }
        right={
          hasCloseButton ? (
            <IconButton onPress={back}>
              <SvgIcon name="close" width={UI_SIZE_24} height={UI_SIZE_24} />
            </IconButton>
          ) : null
        }
      />
      <View style={styles.container}>
        <PasscodeInput
          titleStyleProp={titleStyleProp}
          title={title ?? strings.passcode.input4DigitPasscode}
          description={description ?? strings.passcode.inputPasscodeDesc}
          onComplete={handlePasscodeComplete}
        />
        <PasscodeMessage message={message} messageType={messageType} />
      </View>
    </>
  )
}

export default memo(PasscodeCheckScreen)

const getStyles = createThemedStylesheet((theme) => ({
  container: {
    alignItems: 'center' as 'center',
    backgroundColor: theme.background.bg_1,
    flex: 1,
  },
  title: {
    color: colors.white_snow,
    textAlign: 'center' as 'center',
  },
}))
