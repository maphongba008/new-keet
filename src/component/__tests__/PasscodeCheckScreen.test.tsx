import React from 'react'
import { render } from '@testing-library/react-native'
import { SafeAreaProvider } from 'react-native-safe-area-context'

import PasscodeCheckScreen from 'screen/Passcode/PasscodeCheckScreen'

describe('PasscodeCheckScreen Snapshot', () => {
  it('should match the snapshot with default props', () => {
    const { toJSON } = render(
      <SafeAreaProvider>
        <PasscodeCheckScreen />
      </SafeAreaProvider>,
    )
    expect(toJSON()).toMatchSnapshot()
  })

  it('should match the snapshot with a back button and close button', () => {
    const { toJSON } = render(
      <SafeAreaProvider>
        <PasscodeCheckScreen hasBackButton={true} hasCloseButton={true} />
      </SafeAreaProvider>,
    )
    expect(toJSON()).toMatchSnapshot()
  })

  it('should match the snapshot when onSuccess callback is passed', () => {
    const { toJSON } = render(
      <SafeAreaProvider>
        <PasscodeCheckScreen onSuccess={jest.fn()} />
      </SafeAreaProvider>,
    )
    expect(toJSON()).toMatchSnapshot()
  })
})
