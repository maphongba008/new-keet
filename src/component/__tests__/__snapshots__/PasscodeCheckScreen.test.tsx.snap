// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasscodeCheckScreen Snapshot should match the snapshot when onSuccess callback is passed 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
/>
`;

exports[`PasscodeCheckScreen Snapshot should match the snapshot with a back button and close button 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
/>
`;

exports[`PasscodeCheckScreen Snapshot should match the snapshot with default props 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
/>
`;
